### PR TITLE
Enable to select the metric at the configuration level

### PR DESCRIPTION
**Pull request by sungmanc**
### Notes,
Still, WIP for the pre-commit, unit, and integration tests. However, if you have opinions related to the design change, please feel free to leave your comments.

### This PR introduces,
* Users can select the metrics which are inherited from the torchmetrics at the configuration level.
* Somes metrics that need num_classes information automatically updated by checking the model's `num_classes` information.
* `HLabelAccuracy` doesn't need the complex hlabel information at the `init` level.
* `val_metric` and `test_metric` are integrated --> `metric`.
* All `OTXLitModule` will monitor the learning rate by using the key name of scheduler.

###  TODOs,
* Visualprompting task use the dict format for the metric. However, this PR can't support the dict format. It'll be handled at the later PR.

### Checklist

- [ ] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have added e2e tests for validation.
- [ ] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [ ] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2023 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

File: src/otx/algo/classification/deit_tiny.py
Comment (sungmanc) [inline src/otx/algo/classification/deit_tiny.py:81]: This information is added to use the `Model.num_multiclass_heads` and `Model.num_multilabel_classes` information at the `Metric`.

File: src/otx/core/model/module/base.py
Comment (sungmanc) [inline src/otx/core/model/module/base.py]: scheduler[0] is might `WarmScheduler`, so need to check the `monitor` key of last scheduler.

**Review by goodsong81 — COMMENTED** [review]
Nice to see the metric could be configured, but why don't we enable multiple metrics? The first one might be regarded as the representative metric.

**Comment by sungmanc** [comment]
> Nice to see the metric could be configured, but why don't we enable multiple metrics? The first one might be regarded as the representative metric.

Thanks for your interest, you're right. we'll support the multiple metrics at the next PRs. This is the very first step of giving the freedom of metric-related things. 

Here is the plan, 
1. Enable the chaning the metric at the configuration level
2. Support the previous OTX metric, especially OTX SDK/API metrics
3. Support multiple metric

File: src/otx/cli/cli.py
Comment (harimkang) [inline src/otx/cli/cli.py:300]: As I understand it, `metric` is an argument that depends on `train`, `test` in the subcommand.
model is an argument that belongs to `engine.__init__`, which is fine, but metric might not be. This needs to be checked.

File: src/otx/cli/cli.py
Comment (harimkang) [inline src/otx/cli/cli.py]: It doesn't seem necessary to use so many functions for the metric - couldn't this be done at the same time as updating num_classes in instantiate_model?

File: src/otx/cli/cli.py
Comment (harimkang) [inline src/otx/cli/cli.py:323]: I think this function could also be merged with the `instantiate_model` function, what do you think?

File: src/otx/core/model/module/classification.py
Comment (harimkang) [inline src/otx/core/model/module/classification.py]: In my opinion, `OTXLitModule` should also accept a metric argument.
Also, common code like `metric.reset` should probably go in an `OTXLitModule`.

**Comment by samet-akcay** [comment]
> > Nice to see the metric could be configured, but why don't we enable multiple metrics? The first one might be regarded as the representative metric.
> 
> Thanks for your interest, you're right. we'll support the multiple metrics at the next PRs. This is the very first step of giving the freedom of metric-related things.
> 
> Here is the plan,
> 
> 1. Enable the chaning the metric at the configuration level
> 2. Support the previous OTX metric, especially OTX SDK/API metrics
> 3. Support multiple metric

torchmetrics has a great feature called `MetricCollection` that makes it possible to stack as many metrics you want. LightningModule handles the rest 

https://lightning.ai/docs/torchmetrics/stable/pages/overview.html#metriccollection

File: src/otx/recipe/classification/h_label_cls/efficientnet_b0_light.yaml
Comment (samet-akcay) [inline src/otx/recipe/classification/h_label_cls/efficientnet_b0_light.yaml:22]: I will propose a package hierarchy once the execution is done, but a quick sneak peak. Ideally, it would be easier to access via API if metrics are be located in `otx.metrics`

Comment (sungmanc) [reply src/otx/recipe/classification/h_label_cls/efficientnet_b0_light.yaml:22]: > I will propose a package hierarchy once the execution is done, but a quick sneak peak. Ideally, it would be easier to access via API if metrics are be located in `otx.metrics`

I 100% agree with you. We need to clean the folder architecture

File: src/otx/core/model/module/classification.py
Comment (harimkang) [inline src/otx/core/model/module/classification.py:231]: In near future, we should support auto-configuration for H-Labels as well. To solve this, we need to add logic to automatically set `num_multiclass_heads`, `num_multilabel_classes`. After that logic is added, we can support auto-configuration.

Comment (sungmanc) [reply src/otx/core/model/module/classification.py:231]: I agree, adding the logic of what you said should be done. I think that could be handled at the next PR

**Review by jaegukhyun — DISMISSED** [review]

File: src/otx/core/model/module/action_classification.py
Comment (vinnamkim) [inline src/otx/core/model/module/action_classification.py]: Define `MetricCallable` type such as `LRSchedulerCallable` or `OptimizerCallable`
```suggestion
        metric: list[MetricCallable] | MetricCallable = partial(Accuracy, task="multiclass"),
```

File: src/otx/core/model/module/action_classification.py
Comment (vinnamkim) [inline src/otx/core/model/module/action_classification.py]:
```suggestion
import inspect

...

        sig = inspect.signature()
        if len(sig.parameters) == 0:
             self.metric = metric()
        elif len(sig.parameters) == 1 and "num_classes" in sig.parameters:
             self.metric = metric(num_classes=self.model.num_classes)
        else:
             msg = f"Given metric is not an allowed MetricCallable, {sig}"
             raise ValueError(msg)
```

File: src/otx/core/model/module/base.py
Comment (vinnamkim) [reply src/otx/core/model/module/base.py]: There is no guarantee that `self.scheduler[-1]` has `monitor` attribute and there seems no use of this property after https://github.com/openvinotoolkit/training_extensions/pull/2910. Please just remove this property.

File: src/otx/core/model/module/base.py
Comment (eunwoosh) [inline src/otx/core/model/module/base.py]: It seems that all of tasks initialize `metric` itself again. If then, how about not receiving `metric` from argument and just setting it to `None` in `OTXLitModule`?

File: src/otx/algo/classification/deit_tiny.py
Comment (eunwoosh) [inline src/otx/algo/classification/deit_tiny.py:81]: It seems that all h-label model classes need to save `num_multiclass_heads` and `num_multilabel_classes`, how about passing it to super().__init__() and making it to save them as own attribute?

**Review by eunwoosh — COMMENTED** [review]
Thanks for your work :) I leave some comments. please take a look.

File: src/otx/core/model/module/base.py
Comment (sungmanc) [reply src/otx/core/model/module/base.py]: Thanks for the suggestion, Making the `metric=None` doesn't looks good to me since metric should be delivered (So, can't be None).  As you said, all tasks initialize the `metric` again, I don't want to cause any confusion to the readers that the metric could be None.

File: src/otx/algo/classification/deit_tiny.py
Comment (sungmanc) [reply src/otx/algo/classification/deit_tiny.py:81]: For the `num_multiclass_heads` and `num_multilabel_classes`, it will be removed at the next PR. So, I don't want to enhance the logic at this PR. What do you think?

File: src/otx/recipe/classification/h_label_cls/efficientnet_b0_light.yaml
Comment (sovrasov) [reply src/otx/recipe/classification/h_label_cls/efficientnet_b0_light.yaml:22]: Can we import `HLabelAccuracy` in __init__.py in metrics, and then use `otx.algo.classification.metrics.HLabelAccuracy` to reduce the complexity a bit?

File: src/otx/recipe/classification/h_label_cls/otx_deit_tiny.yaml
Comment (sovrasov) [inline src/otx/recipe/classification/h_label_cls/otx_deit_tiny.yaml:29]: Why do we need to set fictive values here?

File: src/otx/core/model/module/action_classification.py
Comment (sungmanc) [reply src/otx/core/model/module/action_classification.py]: Done, https://github.com/openvinotoolkit/training_extensions/pull/2917/commits/da532871951921f22dc38a1c9b49f1f7e5130deb

File: src/otx/core/model/module/action_classification.py
Comment (sungmanc) [reply src/otx/core/model/module/action_classification.py]: Done, https://github.com/openvinotoolkit/training_extensions/pull/2917/commits/da532871951921f22dc38a1c9b49f1f7e5130deb. Supporting the list of metric will be followed.

File: src/otx/recipe/classification/h_label_cls/otx_deit_tiny.yaml
Comment (sungmanc) [reply src/otx/recipe/classification/h_label_cls/otx_deit_tiny.yaml:29]: `HLabelAccuracy` should take the `num_multiclass_heads` and `num_multilabel_classes`. However, they could be different according to the datasets. So, they will be automatically patched. That's the reason why I just set the default value as 0 for them.

File: src/otx/recipe/classification/h_label_cls/efficientnet_b0_light.yaml
Comment (sungmanc) [reply src/otx/recipe/classification/h_label_cls/efficientnet_b0_light.yaml:22]: I heard that @samet-akcay is planning to the restructuring for the algo folder, it could be handeled at that phase.

**Review by harimkang — APPROVED** [review]

**Review by jaegukhyun — APPROVED** [review]

**Review by eunwoosh — APPROVED** [review]

**Review by sovrasov — APPROVED** [review]

**Review by eugene123tw — APPROVED** [review]

File: src/otx/algo/metrices/__init__.py
Comment (vinnamkim) [inline src/otx/algo/metrices/__init__.py:1]: This should be moved to somewhere in `src/otx/core`

File: src/otx/core/model/module/action_detection.py
Comment (vinnamkim) [inline src/otx/core/model/module/action_detection.py:37]: Why does it cannot have the default MetricCallable and should be `None` as default?

File: src/otx/core/model/module/detection.py
Comment (vinnamkim) [inline src/otx/core/model/module/detection.py:54]: What's the reason that this is created at `on_test_epoch_start()`, not its `__init__()`?

File: src/otx/core/model/module/action_detection.py
Comment (sungmanc) [reply src/otx/core/model/module/action_detection.py:37]: For example, at the export, predict, we don't need add metric.

File: src/otx/core/model/module/detection.py
Comment (sungmanc) [reply src/otx/core/model/module/detection.py:54]: I'm sorry that I can't get your point. I think it's on the __init__ function.

File: src/otx/algo/metrices/__init__.py
Comment (sungmanc) [reply src/otx/algo/metrices/__init__.py:1]: About this, I heard that Samet is designing the new file architecture, so I just want to keep it as-is.

File: src/otx/core/model/module/action_detection.py
Comment (harimkang) [reply src/otx/core/model/module/action_detection.py:37]: It's not necessary for things like exports, but wouldn't it be more convenient for API users to at least have this as the default matric when creating a model?
I think the first case of user scenario is to try training first. However, it would be inconvenient to always include a metric.

**Comment by harimkang** [comment]
We have other workers waiting for this PR, so why don't we merge this for now? Any additional comments should be addressed in the next PR.

**Comment by sungmanc** [comment]
I'll reflect on the remaining reviews after merging this PR.

**Pull request merged**
